## Proyecto:## 

Plan de implementacion para mejorar la asistencia de estudiantes en las provincias de Antofagasta y de Magallanes de Chile

Primero trabajamos dos bases de datos:
-Primero la localización de los colegios de Chile. Que lo bajamos de la página web https://www.kaggle.com/datasets/maxigaarp/localizacion-geografica-de-colegios-de-chile/data. 

-Segundo los datos de la asistencia de los estudiantes de Chile. Que lo bajamos de la página web https://www.kaggle.com/datasets/maxigaarp/asistencia-colegios-2018-y-2019-chile.
Descargamos los archivos csv

In [ ]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("maxigaarp/asistencia-colegios-2018-y-2019-chile")

print("Path to dataset files:", path)

In [ ]:
%pip install kagglehub
import kagglehub

# Download latest version
url= kagglehub.dataset_download("maxigaarp/localizacion-geografica-de-colegios-de-chile")
url2= kagglehub.dataset_download("maxigaarp/asistencia-colegios-2018-y-2019-chile")
print("Path to dataset files:", url)
print("Path to dataset files:", url2)


[notice] A new release of pip is available: 24.3.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.
Path to dataset files: C:\Users\tachi\.cache\kagglehub\datasets\maxigaarp\localizacion-geografica-de-colegios-de-chile\versions\1


Los guardamos en la carpeta del proyecto.

Comenzamos con la creación de la base de datos de localización.

In [61]:
import pandas as pd

df = pd.read_csv("colesgeo.csv") 

In [62]:
df.head()

Unnamed: 0  RBD                              NOM_RBD NOM_COM_RBD  \
0           1    1              LICEO POLITECNICO ARICA       ARICA   
1           2    3  ESC. PEDRO VICENTE GUTIERREZ TORRES       ARICA   
2           3    4            LICEO OCTAVIO PALMA PEREZ       ARICA   
3           4    5             JOVINA NARANJO FERNANDEZ       ARICA   
4           5    7   L. POLI. ANTONIO VARAS DE LA BARRA       ARICA   

                                          query        lon        lat  
0              LICEO POLITECNICO ARICA ,  ARICA -70.294692 -18.487186  
1  ESC. PEDRO VICENTE GUTIERREZ TORRES ,  ARICA -70.302775 -18.486958  
2            LICEO OCTAVIO PALMA PEREZ ,  ARICA -70.307920 -18.474170  
3             JOVINA NARANJO FERNANDEZ ,  ARICA -70.313452 -18.476804  
4   L. POLI. ANTONIO VARAS DE LA BARRA ,  ARICA -70.287950 -18.475946

Haremos un date pequeño de trabajo

anexo: si presentan problemas con la base de datos genere estos dos comandos de código generará un archivo pkl para trabajar

In [ ]:
# df.to_pickle('data.pkl')

In [ ]:
#df = pd.read_pickle('data.pkl')

In [65]:
df.head()

Unnamed: 0  RBD                              NOM_RBD NOM_COM_RBD  \
0           1    1              LICEO POLITECNICO ARICA       ARICA   
1           2    3  ESC. PEDRO VICENTE GUTIERREZ TORRES       ARICA   
2           3    4            LICEO OCTAVIO PALMA PEREZ       ARICA   
3           4    5             JOVINA NARANJO FERNANDEZ       ARICA   
4           5    7   L. POLI. ANTONIO VARAS DE LA BARRA       ARICA   

                                          query        lon        lat  
0              LICEO POLITECNICO ARICA ,  ARICA -70.294692 -18.487186  
1  ESC. PEDRO VICENTE GUTIERREZ TORRES ,  ARICA -70.302775 -18.486958  
2            LICEO OCTAVIO PALMA PEREZ ,  ARICA -70.307920 -18.474170  
3             JOVINA NARANJO FERNANDEZ ,  ARICA -70.313452 -18.476804  
4   L. POLI. ANTONIO VARAS DE LA BARRA ,  ARICA -70.287950 -18.475946

Usaremos la función informe para generar un analisis exploratorio de la base de datos.

In [66]:
def informe(x):
    print("Información del DataFrame:")
    print(x.info())
    print("Descripción estadística:")
    print(x.describe())
    print("Valores nulos:")
    print(x.isnull().sum())
    print("Tipos de datos:")
    print(x.dtypes)
    
    print("Primeras 5 filas:")
    print(x.head())
    print("Últimas 5 filas:")
    print(x.tail())
    print("Forma del DataFrame:")
    print(x.shape)
    print("columnas")
    print(x.columns)
    print("numericos")
    print(x.select_dtypes(include=["number"]).columns)

In [67]:
informe(df)

Información del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9273 entries, 0 to 9272
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Unnamed: 0   9273 non-null   int64  
 1   RBD          9273 non-null   int64  
 2   NOM_RBD      9273 non-null   object 
 3   NOM_COM_RBD  9273 non-null   object 
 4   query        9273 non-null   object 
 5   lon          9178 non-null   float64
 6   lat          9178 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 507.2+ KB
None
Descripción estadística:
        Unnamed: 0           RBD          lon          lat
count  9273.000000   9273.000000  9178.000000  9178.000000
mean   4637.000000  10006.886768   -72.598910   -31.787801
std    2677.028857   8432.644987     6.938825    15.493470
min       1.000000      1.000000  -124.195375   -62.200135
25%    2319.000000   3640.000000   -72.761797   -37.956585
50%    4637.000000   7752.000000   -71.52

Vemos que los datos tienen valores nulos en la columna "latitud" y "longitud". Que posee 7 columnas y 9273 filas de datos.
Elimanaremos las columnas que no necesitamos para nuetsro análisis solo las que serán necesarias para el estudio

In [ ]:
df1=df[["RBD", "NOM_RBD","NOM_COM_RBD", "lon", "lat"]].rename(columns=str.lower)

In [69]:
df1.head()

RBD                              NOM_RBD NOM_COM_RBD        lon        lat
0    1              LICEO POLITECNICO ARICA       ARICA -70.294692 -18.487186
1    3  ESC. PEDRO VICENTE GUTIERREZ TORRES       ARICA -70.302775 -18.486958
2    4            LICEO OCTAVIO PALMA PEREZ       ARICA -70.307920 -18.474170
3    5             JOVINA NARANJO FERNANDEZ       ARICA -70.313452 -18.476804
4    7   L. POLI. ANTONIO VARAS DE LA BARRA       ARICA -70.287950 -18.475946

Revisaremos nuevamente duplicados y nulos a través de la función datos que hemos creado

In [70]:
def datos(x):
    print("datos nulos")
    print(x.isnull().sum())
    print("datos totales")
    print(x.count())   
    print("porcentaje de nulos")
    print(x.isnull().sum() / x.shape[0]*100)
    print("datos duplicados")
    print(x.duplicated().sum())

In [71]:
datos(df1)

datos nulos
RBD             0
NOM_RBD         0
NOM_COM_RBD     0
lon            95
lat            95
dtype: int64
datos totales
RBD            9273
NOM_RBD        9273
NOM_COM_RBD    9273
lon            9178
lat            9178
dtype: int64
porcentaje de nulos
RBD            0.00000
NOM_RBD        0.00000
NOM_COM_RBD    0.00000
lon            1.02448
lat            1.02448
dtype: float64
datos duplicados
0


veremos los datos nulos de las columnas lon y lat que son 95

In [72]:
df_nulos = df1[df1.isnull().any(axis=1)]
df_nulos

RBD                                       NOM_RBD       NOM_COM_RBD  \
211     354                             COL. KID'S SCHOOL       ANTOFAGASTA   
214     357                                 COL. SAN LUIS       ANTOFAGASTA   
434     674                        ESCUELA DE VINITA ALTA            VICU�A   
477     730       ESCUELA BASICA LUCIA NUNEZ DE LA CUADRA            OVALLE   
599     902                ESCUELA BASICA ELENA CAFFARENA            OVALLE   
...     ...                                           ...               ...   
8921  26145               ESC. BAS. LA CANTERA DE MAPUHUE        LA PINTANA   
9027  31005  COL. EDUC VICTOR MANUEL GIOVANETTI ASTUDILLO          LA UNION   
9090  31220             COLEGIO DE ADULTOS VALLE DE ELQUI            VICU�A   
9111  31287              ESCUELA DEL CARI�O RICARDO NAVIA  ESTACION CENTRAL   
9220  40282      COMUNIDAD DE APRENDIZAJE VINCULOS CASTRO            CASTRO   

      lon  lat  
211   NaN  NaN  
214   NaN  NaN  
434   NaN  NaN  
477   NaN  NaN  
599   NaN  NaN  
...   ...  ...  
8921  NaN  NaN  
9027  NaN  NaN  
9090  NaN  NaN  
9111  NaN  NaN  
9220  NaN  NaN  

[95 rows x 5 columns]

Anularemos los nulos de la columna lon y lat con la función limpiar_nulos que nos entregara el dataframe listo para unirlo  a los demás dataframes

In [ ]:
def limpiar_nulos(x):
    df = x.copy()
    df = x.dropna()
    return df

In [74]:
df_final = limpiar_nulos(df1)
df_final

RBD                                          NOM_RBD NOM_COM_RBD  \
0         1                          LICEO POLITECNICO ARICA       ARICA   
1         3              ESC. PEDRO VICENTE GUTIERREZ TORRES       ARICA   
2         4                        LICEO OCTAVIO PALMA PEREZ       ARICA   
3         5                         JOVINA NARANJO FERNANDEZ       ARICA   
4         7               L. POLI. ANTONIO VARAS DE LA BARRA       ARICA   
...     ...                                              ...         ...   
9268  40436  COLEGIO DE LAS ARTES Y LETRAS BETHEL DE QUILPUE     QUILPUE   
9269  40447     CENTRO DE EDUCACION INTEGRADA DE ADULTOS AEK     CALBUCO   
9270  40455                        COLEGIO PATAGONIA INSULAR      CASTRO   
9271  40457           LICEO TECNICO PROFESIONAL PABLO NERUDA      CASTRO   
9272  40459                      CENTRO EDUCACIONAL ALBORADA       ANCUD   

            lon        lat  
0    -70.294692 -18.487186  
1    -70.302775 -18.486958  
2    -70.307920 -18.474170  
3    -70.313452 -18.476804  
4    -70.287950 -18.475946  
...         ...        ...  
9268 -71.435182 -33.048185  
9269 -73.132709 -41.772016  
9270 -73.773764 -42.480971  
9271 -73.773523 -42.482845  
9272 -73.822392 -41.864684  

[9178 rows x 5 columns]

Finalmente, verificaremos que no existan más nulos en el dataframe final con la función de datos

In [75]:
datos(df_final)

datos nulos
RBD            0
NOM_RBD        0
NOM_COM_RBD    0
lon            0
lat            0
dtype: int64
datos totales
RBD            9178
NOM_RBD        9178
NOM_COM_RBD    9178
lon            9178
lat            9178
dtype: int64
porcentaje de nulos
RBD            0.0
NOM_RBD        0.0
NOM_COM_RBD    0.0
lon            0.0
lat            0.0
dtype: float64
datos duplicados
0


listo por último haremos un análisis por columna para verificar que los datos sean correctos. Trabajamos con la función analisis

In [76]:
def analisis(df,col):
    print(df[col].value_counts())
    print(df[col].unique())
    print(df[col].nunique())

Primero la columna rbd:

In [ ]:
analisis(df_final, 'rbd')

RBD
1        1
3        1
4        1
5        1
7        1
        ..
40436    1
40447    1
40455    1
40457    1
40459    1
Name: count, Length: 9178, dtype: int64
[    1     3     4 ... 40455 40457 40459]
9178


Todo bien vamos con las siguientes columnas que es nombre del colegio y nombre de la comuna del colegio:

In [ ]:
analisis(df_final, 'nom_rbd')

NOM_RBD
ESCUELA GABRIELA MISTRAL                         18
ESCUELA MUNICIPAL                                16
ESCUELA ESPANA                                   11
COLEGIO SAN JOSE                                  8
ESCUELA PABLO NERUDA                              8
                                                 ..
ESCUELA HOSPITALARIA ALIANZA                      1
SCUOLA ITALIANA ARTURO DELL� ORO VI�A DEL MAR     1
COLEGIO JOHN WALL HOLCOMB                         1
COLEGIO NAZARET                                   1
COLEGIO INTEGRADO EDUARDO FREI MONTALVA           1
Name: count, Length: 8409, dtype: int64
['LICEO POLITECNICO ARICA' 'ESC. PEDRO VICENTE GUTIERREZ TORRES'
 'LICEO OCTAVIO PALMA PEREZ' ...
 'CENTRO DE EDUCACION INTEGRADA DE ADULTOS AEK'
 'COLEGIO PATAGONIA INSULAR' 'LICEO TECNICO PROFESIONAL PABLO NERUDA']
8409


In [ ]:
analisis(df_final,"nom_com_rbd")

NOM_COM_RBD
TEMUCO            147
LA FLORIDA        139
VI�A DEL MAR      134
VALPARAISO        133
PUENTE ALTO       130
                 ... 
TORTEL              1
PRIMAVERA           1
O'HIGGINS           1
JUAN FERNANDEZ      1
ANTARTICA           1
Name: count, Length: 346, dtype: int64
['ARICA' 'CAMARONES' 'PUTRE' 'GENERAL LAGOS' 'IQUIQUE' 'HUARA' 'CAMI�A'
 'COLCHANE' 'PICA' 'POZO ALMONTE' 'TOCOPILLA' 'MARIA ELENA' 'CALAMA'
 'OLLAG�E' 'SAN PEDRO DE ATACAMA' 'ANTOFAGASTA' 'MEJILLONES'
 'SIERRA GORDA' 'TALTAL' 'CHA�ARAL' 'DIEGO DE ALMAGRO' 'COPIAPO' 'CALDERA'
 'TIERRA AMARILLA' 'VALLENAR' 'FREIRINA' 'HUASCO' 'ALTO DEL CARMEN'
 'LA SERENA' 'LA HIGUERA' 'COQUIMBO' 'ANDACOLLO' 'VICU�A' 'PAIGUANO'
 'OVALLE' 'RIO HURTADO' 'MONTE PATRIA' 'COMBARBALA' 'PUNITAQUI' 'ILLAPEL'
 'SALAMANCA' 'LOS VILOS' 'CANELA' 'LA LIGUA' 'PETORCA' 'CABILDO'
 'ZAPALLAR' 'PAPUDO' 'LOS ANDES' 'SAN FELIPE' 'SAN ESTEBAN' 'CALLE LARGA'
 'RINCONADA' 'PUTAENDO' 'CALERA' 'SANTA MARIA' 'PANQUEHUE' 'LLAILLAY'
 'CATEMU' 

ahora la columna longitudes: lon y lat

In [80]:
analisis(df_final, 'lon')

lon
-84.179968    121
-72.337009     15
-73.479611     15
-73.162731     15
-71.033539     15
             ... 
-70.317369      1
-70.287950      1
-70.313452      1
-70.307920      1
-71.342952      1
Name: count, Length: 7665, dtype: int64
[-70.294692  -70.3027753 -70.3079199 ... -73.7737642 -73.773523
 -73.8223922]
7665


In [81]:
analisis(df_final, 'lat')

lat
 39.739300    121
-34.694925     15
-38.711910     15
-42.168127     15
-39.641600     15
             ... 
-33.047275      1
-33.048185      1
-42.480971      1
-42.482845      1
-41.864684      1
Name: count, Length: 7664, dtype: int64
[-18.4871859 -18.4869579 -18.4741696 ... -42.4809712 -42.482845
 -41.8646841]
7664


Investigando en https://www.bcn.cl/siit/nuestropais/
La República de Chile se sitúa en el margen suroccidental de América del Sur, entre los 17º 30' y 56º 30' de latitud sur.

y se podrian agrupar de la siguiente manera:

-17° a -30° S: Norte de Chile (Arica, Iquique, Antofagasta).
-33° S: Zona central (Santiago/Valparaíso).
-41° a -56° S: Zona sur y austral (Puerto Montt hasta Cabo de Hornos).

Longitud: Chile continental se mueve mayoritariamente entre los -70° y -75° de longitud oeste.

In [ ]:
def clasificar_zona(lat):
    if lat >= -26: return 'Norte Grande'
    elif lat >= -32: return 'Norte Chico'
    elif lat >= -36: return 'Zona Central'
    elif lat >= -41: return 'Zona Sur'
    else: return 'Zona Austral'

df_final['zona'] = df_final['lat'].apply(clasificar_zona)

C:\Users\tachi\AppData\Local\Temp\ipykernel_13500\926185782.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final['Zona'] = df_final['lat'].apply(clasificar_zona)


In [83]:
df_final.head()

RBD                              NOM_RBD NOM_COM_RBD        lon        lat  \
0    1              LICEO POLITECNICO ARICA       ARICA -70.294692 -18.487186   
1    3  ESC. PEDRO VICENTE GUTIERREZ TORRES       ARICA -70.302775 -18.486958   
2    4            LICEO OCTAVIO PALMA PEREZ       ARICA -70.307920 -18.474170   
3    5             JOVINA NARANJO FERNANDEZ       ARICA -70.313452 -18.476804   
4    7   L. POLI. ANTONIO VARAS DE LA BARRA       ARICA -70.287950 -18.475946   

           Zona  
0  Norte Grande  
1  Norte Grande  
2  Norte Grande  
3  Norte Grande  
4  Norte Grande

Generamos el data set de trabajo en un archivo csv nuevo 

In [84]:
df_final.to_csv('localizacion.csv', index=False)

## Veamos los siguientes Data frame ##

Veamos los datos del mes de agosto

In [ ]:
df_agosto = pd.read_csv("082019.csv", sep=";")

In [ ]:
df_agosto.head()

Veamos los datos del mes de junio

In [ ]:
df_junio = pd.read_csv("062019.csv", sep=";")

In [ ]:
df_junio.head()

Analizamos las columnas 

In [ ]:
df_junio.columns

Unimos los datos de los dos archivos 

In [ ]:
df_invierno = pd.concat([df_junio, df_agosto], axis=0, ignore_index = True)

In [ ]:
df_invierno.head()

In [ ]:
df_invierno.tail()

Analizamos las columnas que tiene el dataset

In [ ]:
df_invierno.columns

Cambiamos el nombre de la columna año para tener mayor claridad

In [ ]:
df_invierno.rename(columns= {'AGNO': 'año'}, inplace=True)
df_invierno.columns

Cambiamos el formato de los nombres de las columnas para hacer mas sencillo la referencia.

In [ ]:
df_invierno.columns = df_invierno.columns.str.lower().str.replace(" ", "_")
df_invierno.columns

In [ ]:
df_invierno.head()

In [ ]:
pd.set_option('display.max_columns', None)

### FILTRAMOS COLUMNAS

Función para eliminar columnas que no usaremos

In [ ]:
def drop_columns (df, columns):
    df.drop(columns, axis=1, inplace=True)

In [ ]:
columns_drop = ["rbd", "dgv_rbd", "cod_pro_rbd", "nom_reg_rbd_a", "cod_pro_rbd", "cod_com_rbd", "cod_deprov_rbd", "cod_depe", "cod_depe2", "cod_ense", "cod_ense2", "let_cur", "mrun", "cod_com_alu"]

drop_columns(df_invierno, columns_drop)

In [ ]:
columns_drop = ["cod_reg_rbd", "cod_grado", "fec_nac_alu", "nom_com_alu"]

drop_columns(df_invierno, columns_drop)

In [ ]:
columns_drop = ["asis_promedio"]

drop_columns(df_invierno, columns_drop)

Quedando el siguiente data set:

In [ ]:
df_invierno.head()

### FILTRAMOS POR PROVINCIAS

In [ ]:
df_invierno["nom_deprov_rbd"].value_counts()

Función para limpiar y dejar solo las provincias de interés Magallanes y Antofagasta

In [ ]:
def clean_column(df, column_name):
    df_clean = df.copy()
    df_clean = df_clean[df_clean[column_name].isin(["ANTOFAGASTA - TOCOPILLA", "MAGALLANES"])]
    print(f"Valores únicos en '{column_name}':", 
          df_clean[column_name].unique())
    return df_clean

In [ ]:
df_prov = clean_column(df_invierno, "nom_deprov_rbd")

Finalmente tenemos el dataset:

In [ ]:
df_prov.head()

In [ ]:
df_prov["nom_deprov_rbd"].value_counts()

Hacemos el nuevo csv de datos

In [ ]:
df_prov.to_csv('asistencia.csv', index=False)

### Unamos los datos a través de Merge

haremos la función a nuestros nuevos archivos de datos.

In [ ]:
df_locaciones= pd.read_csv("localizacion.csv") 
df_asistencia= pd.read_csv("asistencia.csv")


In [ ]:
def union(df1,df2):
    df_union=pd.merge(df1,df2,on=["nom_rbd","nom_com_rbd"],how='left')
    return df_union

In [ ]:
df_union = union(df_asistencia,df_locaciones)   
df_union.head()

Hemos creado un data set con ambos  data frames

Veamos el analisis de nuestro data frame 

In [ ]:
informe(df_union)

Observamos que tenemos datos nulos en algunas columnas haremos un analisis de estos datos usando nuestras funciones

In [ ]:
datos(df_union)

Observemos en detalle algunos de los datos nulos

In [ ]:
df_nulos = df_union[df_union.isnull().any(axis=1)]
df_nulos

Limpiaremos con esto 30.137 datos que les falta información haciendo la limpieza de datos con funciones

In [ ]:
df_union_limpio = limpiar_nulos(df_union)

In [ ]:
df_union_limpio

Lista la base de estudio con 209412 filas y 13 columnas. Generaremos un nuevo csv con los datos limpios

In [ ]:
df_union_limpio.to_csv('datos_finales.csv', index=False)